test/e2e/gateway/gateway_test.go

```diff
@@ -19,134 +19,150 @@ import (
 	"context"
 	"testing"
 
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
 	"github.com/projectcontour/contour/test/e2e"
-	"github.com/stretchr/testify/require"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	gatewayv1alpha1 "sigs.k8s.io/gateway-api/apis/v1alpha1"
 )
 
-type testGroup struct {
-	name    string
-	gateway *gatewayv1alpha1.Gateway
-	tests   map[string]func(t *testing.T, f *e2e.Framework)
+func TestGatewayAPI(t *testing.T) {
+	RegisterFailHandler(Fail)
+	RunSpecs(t, "Gateway API tests")
 }
 
-// getTests defines the test groups that make up the Gateway test suite.
-// Each group shares a single Gateway definition.
-func getTests() []*testGroup {
-	var testGroups []*testGroup
-
-	testGroups = append(testGroups, &testGroup{
-		name: "default gateway",
-		tests: map[string]func(t *testing.T, f *e2e.Framework){
-			"001-path-condition-match":               testGatewayPathConditionMatch,
-			"002-header-condition-match":             testGatewayHeaderConditionMatch,
-			"003-invalid-forward-to":                 testInvalidForwardTo,
-			"005-request-header-modifier-forward-to": testRequestHeaderModifierForwardTo,
-			"005-request-header-modifier-rule":       testRequestHeaderModifierRule,
-		},
-		gateway: &gatewayv1alpha1.Gateway{
-			// Namespace and name need to match what's
-			// configured in the Contour config file.
-			ObjectMeta: metav1.ObjectMeta{
-				Namespace: "projectcontour",
-				Name:      "contour",
-			},
-			Spec: gatewayv1alpha1.GatewaySpec{
-				GatewayClassName: "contour-class",
-				Listeners: []gatewayv1alpha1.Listener{
-					{
-						Protocol: gatewayv1alpha1.HTTPProtocolType,
-						Port:     gatewayv1alpha1.PortNumber(80),
-						Routes: gatewayv1alpha1.RouteBindingSelector{
-							Kind: "HTTPRoute",
-							Namespaces: &gatewayv1alpha1.RouteNamespaces{
-								From: routeSelectTypePtr(gatewayv1alpha1.RouteSelectAll),
-							},
-							Selector: &metav1.LabelSelector{
-								MatchLabels: map[string]string{"app": "filter"},
+var _ = Describe("Gateway API", func() {
+	var f *e2e.Framework
+
+	BeforeEach(func() {
+		f = e2e.NewFramework(GinkgoT())
+	})
+
+	Describe("Insecure (Non-TLS) Gateway", func() {
+		var gateway *gatewayv1alpha1.Gateway
+
+		BeforeEach(func() {
+			gateway = &gatewayv1alpha1.Gateway{
```

Comment:
It doesn't really make sense to have this in a `BeforeEach`, I should just need to create the Gateway once for the subsequent set of specs. Deleting it and re-creating it for each spec isn't a problem, it's just extraneous operations. I guess my alternative is to have each unique Gateway config be its own test suite (file), and then use `BeforeSuite` in each to set up the gateway?

Reply:
we could also use `Context` or `Describe` and segment unique gateways in that, maybe the before/aftereach each does an "ensure" instead of create/delete?

Reply:
ha, didnt really read the code before this comment, disregard, thats what you already have ^

Reply:
Yeah, I think it's fine for now but potentially something to refine. I'll add a code comment about it.